Client/WarFare/UIItemUpgrade.cpp

```diff
@@ -0,0 +1,1289 @@
+				{
+					pStr->SetVisible(false);
+				}
+				else
+				{
+					if (m_pMyUpgradeInv[i]->pUIIcon->IsVisible())
+					{
+						pStr->SetVisible(true);
+						pStr->SetStringAsInt(m_pMyUpgradeInv[i]->iCount);
+						pStr->Render();
+					}
+					else
+					{
+						pStr->SetVisible(false);
+					}
+				}
+			}
+		}
+		else
+		{
+			char szID[32];
+			sprintf(szID, "s_count_%d", i);
+			CN3UIString* pStr = (CN3UIString*)GetChildByID(szID);
+			if (pStr)
+				pStr->SetVisible(false);
+		}
+	}
+
+	if ((GetState() == UI_STATE_ICON_MOVING) && (CN3UIWndBase::m_sSelectedIconInfo.pItemSelect))
+		CN3UIWndBase::m_sSelectedIconInfo.pItemSelect->pUIIcon->Render();
+
+	if (spItem)
+	{
+		m_pUITooltipDlg->DisplayTooltipsEnable(ptCur.x, ptCur.y, spItem, false, false);
+	}
+}
+
+void CUIItemUpgrade::InitIconWnd(e_UIWND eWnd)
+{
+	__TABLE_UI_RESRC* pTbl = CGameBase::s_pTbl_UI.Find(CGameBase::s_pPlayer->m_InfoBase.eNation);
+
+	m_pUITooltipDlg = new CUIImageTooltipDlg();
+	m_pUITooltipDlg->Init(this);
+	m_pUITooltipDlg->LoadFromFile(pTbl->szItemInfo);
+	m_pUITooltipDlg->InitPos();
+	m_pUITooltipDlg->SetVisible(FALSE);
+
+	CN3UIWndBase::InitIconWnd(eWnd);
+
+	m_pStrMyGold = (CN3UIString*) GetChildByID("text_gold");
+	__ASSERT(m_pStrMyGold, "NULL UI Component!!");
+	if (m_pStrMyGold) m_pStrMyGold->SetString("0");
+}
+
+__IconItemSkill* CUIItemUpgrade::GetHighlightIconItem(CN3UIIcon* pUIIcon)
+{
+	int i;
+	for (i = 0; i < MAX_ITEM_UPGRADE_SLOT; i++)
+	{
+		if ((m_pMyUpgradeSLot[i] != nullptr) && (m_pMyUpgradeSLot[i]->pUIIcon == pUIIcon))
+			return m_pMyUpgradeSLot[i];
+	}
+
+	for (i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if ((m_pMyUpgradeInv[i] != nullptr) && (m_pMyUpgradeInv[i]->pUIIcon == pUIIcon))
+			return m_pMyUpgradeInv[i];
+	}
+	if (m_pUpgradeItemSlot && m_pUpgradeItemSlot->pUIIcon == pUIIcon)
+		return m_pUpgradeItemSlot;
+
+	return nullptr;
+}
+
+void CUIItemUpgrade::Open()
+{
+	SetVisible(true);
+	ItemMoveFromInvToThis();
+
+	GoldUpdate();
+}
+
+void CUIItemUpgrade::GoldUpdate()
+{
+	if (m_pStrMyGold)
+	{
+		m_pStrMyGold->SetStringAsInt(CGameBase::s_pPlayer->m_InfoExt.iGold);
+	}
+}
+
+void CUIItemUpgrade::ItemMoveFromInvToThis()
+{
+	CUIInventory* pInven = CGameProcedure::s_pProcMain->m_pUIInventory;
+	if (!pInven)
+		return;
+
+
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if (pInven->m_pMyInvWnd[i])
+		{
+			__IconItemSkill* spItem = pInven->m_pMyInvWnd[i];
+			spItem->pUIIcon->SetParent(this);
+
+			pInven->m_pMyInvWnd[i] = nullptr;
+			CN3UIArea* pArea;
+
+			char szID[32];
+			sprintf(szID, "a_slot_%d", i);
+			pArea = (CN3UIArea*)GetChildByID(szID);
+			if (pArea)
+			{
+				spItem->pUIIcon->SetRegion(pArea->GetRegion());
+				spItem->pUIIcon->SetMoveRect(pArea->GetRegion());
+			}
+
+			m_pMyUpgradeInv[i] = spItem;
+		}
+		// Backup the inventory state for restoration if needed.
+		m_pBackupUpgradeInv[i] = nullptr;
+		if (m_pMyUpgradeInv[i])
+		{
+			m_pBackupUpgradeInv[i] = new __IconItemSkill(*m_pMyUpgradeInv[i]);
```

Comment:
This is very dangerous. We don't define a copy operator for this.
As-is, this will copy pointers as-is (e.g. for the icon), rather than copying the icons themselves.

More than this though is the logic in general; managing 2 separate arrays of item data, one of them the inventory's real data, and then a copy for "restoration purposes" just seems painful to maintain.
Just simplify it by only copying from the inventory (don't steal its icons), and syncing the relevant change later on the upgrade packet.

Then the items here don't matter in the slightest and won't affect anything.